base/compiler/abstractinterpretation.jl

```diff
@@ -640,8 +671,8 @@ end
 
 function pure_eval_eligible(interp::AbstractInterpreter,
     @nospecialize(f), applicable::Vector{Any}, arginfo::ArgInfo, sv::InferenceState)
-    return !isoverlayed(method_table(interp)) &&
-           f !== nothing &&
+    # XXX we need to check that this pure function doesn't call any overlayed method
```

Comment:
For correctness, we need to look at the effect of the pure function being analyzed here and assert that it doesn't involve any overlayed method call within it.
But I think a pure function isn't really supposed to call a generic function (which may be overlayed) anyway, and so here I'd like to avoid trying hard to exclude such rare and invalid usecases and adding implementation complexity